File: libpretixsync/src/main/java/eu/pretix/libpretixsync/api/PretixApi.java
package eu.pretix.libpretixsync.api;

import eu.pretix.libpretixsync.DummySentryImplementation;
import eu.pretix.libpretixsync.SentryInterface;
import eu.pretix.libpretixsync.check.TicketCheckProvider;
import eu.pretix.libpretixsync.config.ConfigStore;
import eu.pretix.libpretixsync.db.QueuedCheckIn;
import eu.pretix.libpretixsync.utils.NetUtils;
import okhttp3.*;

import org.json.JSONException;
import org.json.JSONObject;

import javax.net.ssl.SSLException;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLEncoder;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.TimeZone;

public class PretixApi {
    /**
     * See https://docs.pretix.eu/en/latest/api/index.html for API documentation
     */

    public static final int SUPPORTED_API_VERSION = 4;
    public static final MediaType JSON
            = MediaType.parse("application/json; charset=utf-8");

    private String url;
    private String eventSlug;
    private String orgaSlug;
    private String key;
    private int version;
    private OkHttpClient client;
    private SentryInterface sentry;

    public class ApiResponse {
        private JSONObject data;
        private Response response;

        public ApiResponse(JSONObject data, Response response) {
            this.data = data;
            this.response = response;
        }

        public JSONObject getData() {
            return data;
        }

        public Response getResponse() {
            return response;
        }
    }

    public PretixApi(String url, String key, String orgaSlug, String eventSlug, int version, HttpClientFactory httpClientFactory) {
        if (!url.endsWith("/")) {
            url += "/";
        }
        this.url = url;
        this.key = key;
        this.eventSlug = eventSlug;
        this.orgaSlug = orgaSlug;
        this.version = version;
        this.client = httpClientFactory.buildClient(NetUtils.ignoreSSLforURL(url));
        this.sentry = new DummySentryImplementation();
    }

    public String getEventSlug() {
        return eventSlug;
    }

    public void setEventSlug(String eventSlug) {
        this.eventSlug = eventSlug;
    }

    public static PretixApi fromConfig(ConfigStore config, HttpClientFactory httpClientFactory) {
        return new PretixApi(config.getApiUrl(), config.getApiKey(), config.getOrganizerSlug(),
                config.getEventSlug(), config.getApiVersion(), httpClientFactory);
    }

    public static PretixApi fromConfig(ConfigStore config) {
        return PretixApi.fromConfig(config, new DefaultHttpClientFactory());
    }

    public ApiResponse redeem(String secret, Date datetime, boolean force, String nonce, List<TicketCheckProvider.Answer> answers, Long listId, boolean ignore_unpaid, boolean pdf_data) throws ApiException, JSONException {
        String dt = null;
        if (datetime != null) {
            dt = QueuedCheckIn.formatDatetime(datetime);
        }
        return redeem(secret, dt, force, nonce, answers, listId, ignore_unpaid, pdf_data);
    }

    public ApiResponse redeem(String secret, String datetime, boolean force, String nonce, List<TicketCheckProvider.Answer> answers, Long listId, boolean ignore_unpaid, boolean pdf_data) throws ApiException, JSONException {
        JSONObject body = new JSONObject();
        if (datetime != null) {
            body.put("datetime", datetime);
        }
        body.put("force", force);
        body.put("ignore_unpaid", ignore_unpaid);
        body.put("nonce", nonce);
        JSONObject answerbody = new JSONObject();
        if (answers != null) {
            for (TicketCheckProvider.Answer a : answers) {
                answerbody.put("" + a.getQuestion().getServer_id(), a.getValue());
            }
        }
        body.put("answers", answerbody);
        body.put("questions_supported", true);
        body.put("canceled_supported", true);
        String pd = "";
        if (pdf_data) {
            pd = "?pdf_data=true";
        }
        return postResource(eventResourceUrl("checkinlists/" + listId + "/positions/" + secret + "/redeem") + pd, body);
    }

    public ApiResponse status(Long listId) throws ApiException {
        try {
            return fetchResource(eventResourceUrl("checkinlists/" + listId + "/status"));
        } catch (ResourceNotModified resourceNotModified) {
            throw new FinalApiException("invalid error");
        }
    }

    public ApiResponse search(Long listId, String query, int page) throws ApiException {
        try {
            return fetchResource(eventResourceUrl("checkinlists/" + listId + "/positions") + "?ignore_status=true&page=" + page + "&search=" + URLEncoder.encode(query, "UTF-8"));
        } catch (ResourceNotModified | UnsupportedEncodingException resourceNotModified) {
            throw new FinalApiException("invalid error");
        }
    }

    public String apiURL(String suffix) {
        try {
            return new URL(new URL(url), "/api/v1/" + suffix).toString();
        } catch (MalformedURLException e) {
            e.printStackTrace();
            return null;
        }
    }

    public String organizerResourceUrl(String resource) {
        try {
            return new URL(new URL(url), "/api/v1/organizers/" + orgaSlug + "/" + resource + "/").toString();
        } catch (MalformedURLException e) {
            e.printStackTrace();
            return null;
        }
    }

    public String eventResourceUrl(String resource) {
        try {
            return new URL(new URL(url), "/api/v1/organizers/" + orgaSlug + "/events/" + eventSlug + "/" + resource + "/").toString();
        } catch (MalformedURLException e) {
            e.printStackTrace();
            return null;
        }
    }

    public ApiResponse deleteResource(String full_url) throws ApiException {
        Request.Builder request = new Request.Builder()
                .url(full_url)
                .delete()
                .header("Authorization", "Device " + key);
        try {
            return apiCall(request.build(), false);
        } catch (ResourceNotModified resourceNotModified) {
            resourceNotModified.printStackTrace();
            throw new FinalApiException("Resource not modified");
        }
    }

    public ApiResponse postResource(String full_url, JSONObject data) throws ApiException {
        return postResource(full_url, data, null);
    }

    public ApiResponse postResource(String full_url, JSONObject data, String idempotency_key) throws ApiException {
        Request.Builder request = new Request.Builder()
                .url(full_url)
                .post(RequestBody.create(data.toString(), MediaType.parse("application/json")))
                .header("Authorization", "Device " + key);
        if (idempotency_key != null) {
            request = request.header("X-Idempotency-Key", idempotency_key);
        }
        try {
            return apiCall(request.build());
        } catch (ResourceNotModified resourceNotModified) {
            resourceNotModified.printStackTrace();
            throw new FinalApiException("Resource not modified");
        }
    }

    public ApiResponse fetchResource(String full_url, String if_modified_since) throws ApiException, ResourceNotModified {
        Request.Builder request = new Request.Builder()
                .url(full_url)
                .header("Authorization", "Device " + key);
        if (if_modified_since != null) {
            request = request.header("If-Modified-Since", if_modified_since);
        }
        return apiCall(request.get().build());
    }

    public ApiResponse fetchResource(String full_url) throws ApiException, ResourceNotModified {
        return fetchResource(full_url, null);
    }

    public ApiResponse downloadFile(String full_url) throws ApiException {
        Request.Builder request = new Request.Builder()
                .url(full_url)
                .header("Authorization", "Device " + key);
        try {
            return apiCall(request.build(), false);
        } catch (ResourceNotModified resourceNotModified) {
            resourceNotModified.printStackTrace();
            throw new FinalApiException("Resource not modified");
        }
    }

    public SentryInterface getSentry() {
        return sentry;
    }

    public void setSentry(SentryInterface sentry) {
        this.sentry = sentry;
    }

    private ApiResponse apiCall(Request request) throws ApiException, ResourceNotModified {
        return apiCall(request, true);
    }

    private ApiResponse apiCall(Request request, boolean json) throws ApiException, ResourceNotModified {
        Response response;
        try {
            response = client.newCall(request).execute();
        } catch (SSLException e) {
            e.printStackTrace();
            throw new ApiException("Error while creating a secure connection.", e);
        } catch (IOException e) {
            e.printStackTrace();
            throw new ApiException("Connection error: " + e.getMessage(), e);
        }

        String safe_url = request.url().toString().replaceAll("^(.*)key=([0-9A-Za-z]+)([^0-9A-Za-z]*)", "$1key=redacted$3");
        sentry.addHttpBreadcrumb(safe_url, request.method(), response.code());
        String body = "";
        if (json) {
            try {
                body = response.body().string();
            } catch (IOException e) {
                e.printStackTrace();
                throw new ApiException("Connection error: " + e.getMessage(), e);
            }
        }

        if (response.code() >= 500) {
            response.close();
            throw new ApiException("Server error: " + response.code());
        } else if (response.code() == 404 && (!json || !body.startsWith("{"))) {
            response.close();
            throw new FinalApiException("Server error: Resource not found.");
        } else if (response.code() == 304) {
            throw new ResourceNotModified();
        } else if (response.code() == 403) {
            response.close();
            throw new FinalApiException("Server error: Permission denied.");
        } else if (response.code() == 409) {
            response.close();
            throw new ConflictApiException("Server error: " + response.code() + ": " + body);
        } else if (response.code() >= 405) {
            response.close();
            throw new FinalApiException("Server error: " + response.code() + ".");
        }
        if (response.code() == 401) {
            if (body.startsWith("{")) {
                try {
                    JSONObject err = new JSONObject(body);
                    if (err.optString("detail", "").equals("Device access has been revoked.")) {
                        throw new DeviceAccessRevokedException("Device access has been revoked.");
                    }
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }
        }
        try {
            if (json) {
                if (body.startsWith("[")) {
                    body = "{\"content\": " + body + "}";
                }
                return new ApiResponse(
                        new JSONObject(body),
                        response
                );
            } else {
                return new ApiResponse(
                        null,
                        response
                );
            }
        } catch (JSONException e) {
            e.printStackTrace();
            sentry.captureException(e);
            throw new ApiException("Invalid JSON received: " + body.substring(0, 100), e);
        }
    }

}

File: libpretixsync/src/main/java/eu/pretix/libpretixsync/api/PretixApi.kt
package eu.pretix.libpretixsync.api

import eu.pretix.libpretixsync.DummySentryImplementation
import eu.pretix.libpretixsync.SentryInterface
import eu.pretix.libpretixsync.config.ConfigStore
import eu.pretix.libpretixsync.db.Answer
import eu.pretix.libpretixsync.db.Question
import eu.pretix.libpretixsync.db.QueuedCheckIn
import eu.pretix.libpretixsync.utils.NetUtils
import eu.pretix.libpretixsync.utils.URLFragmentEncoder
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.File
import java.io.IOException
import java.io.InterruptedIOException
import java.io.UnsupportedEncodingException
import java.net.MalformedURLException
import java.net.SocketTimeoutException
import java.net.URL
import java.net.URLEncoder
import java.nio.charset.Charset
import java.util.*
import java.util.concurrent.TimeUnit
import javax.net.ssl.SSLPeerUnverifiedException

open class PretixApi(url: String, key: String, orgaSlug: String, version: Int, httpClientFactory: HttpClientFactory, val acceptLanguage: String? = null) {
    private val url: String
    private val orgaSlug: String
    private val key: String
    private val version: Int
    private val client: OkHttpClient
    var sentry: SentryInterface

    inner class ApiResponse(val data: JSONObject?, val response: Response)

    @Throws(ApiException::class, JSONException::class)
    fun redeem(eventSlug: String, secret: String, datetime: Date?, force: Boolean, nonce: String?, answers: List<Answer>?, listId: Long, ignore_unpaid: Boolean, pdf_data: Boolean, type: String?, source_type: String?, callTimeout: Long? = null, questions_supported: Boolean = true): ApiResponse {
        var dt: String? = null
        if (datetime != null) {
            dt = QueuedCheckIn.formatDatetime(datetime)
        }
        return redeem(eventSlug, secret, dt, force, nonce, answers, listId, ignore_unpaid, pdf_data, type, source_type, callTimeout, questions_supported)
    }

    @Throws(ApiException::class, JSONException::class)
    open fun redeem(eventSlug: String, secret: String, datetime: String?, force: Boolean, nonce: String?, answers: List<Answer>?, listId: Long, ignore_unpaid: Boolean, pdf_data: Boolean, type: String?, source_type: String?, callTimeout: Long? = null, questions_supported: Boolean=true): ApiResponse {
        val body = JSONObject()
        if (datetime != null) {
            body.put("datetime", datetime)
        }
        body.put("force", force)
        body.put("ignore_unpaid", ignore_unpaid)
        body.put("nonce", nonce)
        body.put("type", type)
        body.put("source_type", source_type ?: "barcode")
        val answerbody = JSONObject()
        if (answers != null) {
            for (a in answers) {
                if (a.value.startsWith("file:///")) {
                    val fileid = uploadFile(File(a.value.substring(7)), when (a.value.split(".").last()) {
                        "jpeg", "jpg" -> "image/jpeg".toMediaTypeOrNull()!!
                        "png" -> "image/png".toMediaTypeOrNull()!!
                        "gif" -> "image/gif".toMediaTypeOrNull()!!
                        "pdf" -> "application/pdf".toMediaTypeOrNull()!!
                        else -> "application/unknown".toMediaTypeOrNull()!!
                    }, a.value.split("/").last())
                    answerbody.put("" + (a.question as Question).getServer_id(), fileid)
                } else {
                    answerbody.put("" + (a.question as Question).getServer_id(), a.value)
                }
            }
        }
        body.put("answers", answerbody)
        body.put("questions_supported", questions_supported)
        body.put("canceled_supported", true)
        var pd = ""
        if (pdf_data) {
            pd = "?pdf_data=true"
        }
        return postResource(eventResourceUrl(eventSlug, "checkinlists/" + listId + "/positions/" + URLFragmentEncoder.STRICT.encode(secret, Charset.forName("UTF-8")) + "/redeem") + pd, body, null, callTimeout = callTimeout)
    }

    @Throws(ApiException::class, JSONException::class)
    fun redeem(lists: List<Long>, secret: String, datetime: Date?, force: Boolean, nonce: String?, answers: List<Answer>?, ignore_unpaid: Boolean, pdf_data: Boolean, type: String?, callTimeout: Long? = null, questions_supported: Boolean = true): ApiResponse {
        var dt: String? = null
        if (datetime != null) {
            dt = QueuedCheckIn.formatDatetime(datetime)
        }
        return redeem(lists, secret, dt, force, nonce, answers, ignore_unpaid, pdf_data, type, callTimeout, questions_supported)
    }

    @Throws(ApiException::class, JSONException::class)
    open fun redeem(lists: List<Long>, secret: String, datetime: String?, force: Boolean, nonce: String?, answers: List<Answer>?, ignore_unpaid: Boolean, pdf_data: Boolean, type: String?, callTimeout: Long? = null, questions_supported: Boolean = true): ApiResponse {
        val body = JSONObject()
        if (datetime != null) {
            body.put("datetime", datetime)
        }
        body.put("force", force)
        body.put("ignore_unpaid", ignore_unpaid)
        body.put("nonce", nonce)
        body.put("type", type)
        val answerbody = JSONObject()
        if (answers != null) {
            for (a in answers) {
                if (a.value.startsWith("file:///")) {
                    val fileid = uploadFile(File(a.value.substring(7)), when (a.value.split(".").last()) {
                        "jpeg", "jpg" -> "image/jpeg".toMediaTypeOrNull()!!
                        "png" -> "image/png".toMediaTypeOrNull()!!
                        "gif" -> "image/gif".toMediaTypeOrNull()!!
                        "pdf" -> "application/pdf".toMediaTypeOrNull()!!
                        else -> "application/unknown".toMediaTypeOrNull()!!
                    }, a.value.split("/").last())
                    answerbody.put("" + (a.question as Question).getServer_id(), fileid)
                } else {
                    answerbody.put("" + (a.question as Question).getServer_id(), a.value)
                }
            }
        }
        body.put("answers", answerbody)
        body.put("questions_supported", questions_supported)
        body.put("canceled_supported", true)
        body.put("secret", secret)
        val jlists = JSONArray()
        for (l in lists) {
            jlists.put(l)
        }
        body.put("lists", jlists)
        var pd = "?expand=answers.question"
        if (pdf_data) {
            pd += "&pdf_data=true"
        }
        return postResource(organizerResourceUrl("checkinrpc/redeem") + pd, body, null, callTimeout = callTimeout)
    }

    @Throws(ApiException::class)
    open fun status(eventSlug: String, listId: Long): ApiResponse {
        return try {
            fetchResource(eventResourceUrl(eventSlug, "checkinlists/$listId/status"))
        } catch (resourceNotModified: ResourceNotModified) {
            throw FinalApiException("invalid error")
        }
    }

    @Throws(ApiException::class)
    open fun search(eventSlug: String, listId: Long, query: String?, page: Int): ApiResponse {
        return try {
            fetchResource(eventResourceUrl(eventSlug, "checkinlists/$listId/positions") + "?ignore_status=true&page=" + page + "&search=" + URLEncoder.encode(query, "UTF-8"))
        } catch (resourceNotModified: ResourceNotModified) {
            throw FinalApiException("invalid error")
        } catch (resourceNotModified: UnsupportedEncodingException) {
            throw FinalApiException("invalid error")
        }
    }

    @Throws(ApiException::class)
    open fun search(lists: List<Long>, query: String?, page: Int): ApiResponse {
        return try {
            var url = organizerResourceUrl("checkinrpc/search") + "?ignore_status=true&page=" + page + "&search=" + URLEncoder.encode(query, "UTF-8")
            for (l in lists) {
                url += "&list=$l"
            }
            fetchResource(url)
        } catch (resourceNotModified: ResourceNotModified) {
            throw FinalApiException("invalid error")
        } catch (resourceNotModified: UnsupportedEncodingException) {
            throw FinalApiException("invalid error")
        }
    }

    fun apiURL(suffix: String): String {
        return try {
            URL(URL(url), "/api/v1/$suffix").toString()
        } catch (e: MalformedURLException) {
            e.printStackTrace()
            ""
        }
    }

    fun organizerResourceUrl(resource: String): String {
        return try {
            URL(URL(url), "/api/v1/organizers/$orgaSlug/$resource/").toString()
        } catch (e: MalformedURLException) {
            e.printStackTrace()
            ""
        }
    }

    fun eventResourceUrl(eventSlug: String, resource: String): String {
        return try {
            URL(URL(url), "/api/v1/organizers/$orgaSlug/events/${eventSlug}/$resource/").toString()
        } catch (e: MalformedURLException) {
            e.printStackTrace()
            ""
        }
    }

    @Throws(ApiException::class)
    open fun deleteResource(full_url: String, idempotency_key: String? = null): ApiResponse {
        var request = Request.Builder()
                .url(full_url)
                .delete()
                .header("Authorization", "Device $key")
        if (acceptLanguage != null) {
            request = request.header("Accept-Language", acceptLanguage)
        }
        if (idempotency_key != null) {
            request = request.header("X-Idempotency-Key", idempotency_key)
        }
        return try {
            apiCall(request.build(), false)
        } catch (resourceNotModified: ResourceNotModified) {
            resourceNotModified.printStackTrace()
            throw FinalApiException("Resource not modified")
        }
    }

    @Throws(ApiException::class)
    fun patchResource(full_url: String, data: JSONObject): ApiResponse {
        return patchResource(full_url, data, null)
    }

    @Throws(ApiException::class)
    open fun patchResource(full_url: String?, data: JSONObject, idempotency_key: String?, callTimeout: Long? = null): ApiResponse {
        var request = Request.Builder()
                .url(full_url!!)
                .patch(data.toString().toRequestBody("application/json".toMediaType()))
                .header("Authorization", "Device $key")
        if (acceptLanguage != null) {
            request = request.header("Accept-Language", acceptLanguage)
        }
        if (idempotency_key != null) {
            request = request.header("X-Idempotency-Key", idempotency_key)
        }
        return try {
            apiCall(request.build(), callTimeout = callTimeout)
        } catch (resourceNotModified: ResourceNotModified) {
            resourceNotModified.printStackTrace()
            throw FinalApiException("Resource not modified")
        }
    }

    @Throws(ApiException::class)
    fun postResource(full_url: String, data: JSONObject): ApiResponse {
        return postResource(full_url, data.toString(), null)
    }

    @Throws(ApiException::class)
    fun postResource(full_url: String, data: JSONArray): ApiResponse {
        return postResource(full_url, data.toString(), null)
    }

    @Throws(ApiException::class)
    fun postResource(full_url: String, data: JSONObject, idempotency_key: String?, callTimeout: Long? = null): ApiResponse {
        return postResource(full_url, data.toString(), idempotency_key, callTimeout)
    }

    @Throws(ApiException::class)
    fun postResource(full_url: String, data: JSONObject, idempotency_key: String?): ApiResponse {
        return postResource(full_url, data.toString(), idempotency_key, null)
    }

    @Throws(ApiException::class)
    fun postResource(full_url: String, data: String, idempotency_key: String?): ApiResponse {
        return postResource(full_url, data, idempotency_key, null)
    }

    @Throws(ApiException::class)
    open fun postResource(full_url: String, data: String, idempotency_key: String?, callTimeout: Long? = null): ApiResponse {
        var request = Request.Builder()
                .url(full_url)
                .post(data.toRequestBody("application/json".toMediaType()))
                .header("Authorization", "Device $key")
        if (acceptLanguage != null) {
            request = request.header("Accept-Language", acceptLanguage)
        }
        if (idempotency_key != null) {
            request = request.header("X-Idempotency-Key", idempotency_key)
        }
        return try {
            apiCall(request.build(), callTimeout = callTimeout)
        } catch (resourceNotModified: ResourceNotModified) {
            resourceNotModified.printStackTrace()
            throw FinalApiException("Resource not modified")
        }
    }

    @Throws(ApiException::class, ResourceNotModified::class)
    open fun fetchResource(full_url: String, if_modified_since: String? = null): ApiResponse {
        var request = Request.Builder()
                .url(full_url)
                .header("Authorization", "Device $key")
        if (acceptLanguage != null) {
            request = request.header("Accept-Language", acceptLanguage)
        }
        if (if_modified_since != null) {
            request = request.header("If-Modified-Since", if_modified_since)
        }
        return apiCall(request.get().build())
    }

    @Throws(ApiException::class, ResourceNotModified::class)
    fun fetchResource(full_url: String): ApiResponse {
        return fetchResource(full_url, null)
    }

    @Throws(ApiException::class)
    open fun downloadFile(full_url: String): ApiResponse? {
        var request = Request.Builder()
                .url(full_url)
                .header("Authorization", "Device $key")
        if (acceptLanguage != null) {
            request = request.header("Accept-Language", acceptLanguage)
        }
        return try {
            apiCall(request.build(), false)
        } catch (resourceNotModified: ResourceNotModified) {
            resourceNotModified.printStackTrace()
            throw FinalApiException("Resource not modified")
        }
    }

    @Throws(ApiException::class, ResourceNotModified::class)
    private fun apiCall(
        request: Request,
        json: Boolean = true,
        is_retry: Boolean = false,
        callTimeout: Long? = null
    ): ApiResponse {
        val response: Response
        val httpClient = if (callTimeout != null) {
            client.newBuilder()
                .connectTimeout(callTimeout, TimeUnit.MILLISECONDS)
                .readTimeout(callTimeout, TimeUnit.MILLISECONDS)
                .writeTimeout(callTimeout, TimeUnit.MILLISECONDS)
                .callTimeout(callTimeout, TimeUnit.MILLISECONDS)
                .build()
        } else {
            client
        }
        response = try {
            httpClient.newCall(request).execute()
        } catch (e: SSLPeerUnverifiedException) {
            if (!is_retry) {
                // On Windows, we occasionally see SSL errors after a long period of idle. We didn't fully figure
                // it out, but it seems to be because it tries to reuse a SSL socket that the server already closed.
                // We assume it's safe to retry in all cases since the verification happens before any payload
                // reaches the server.
                httpClient.connectionPool.evictAll()
                return apiCall(request, json, is_retry, callTimeout)
            }
            e.printStackTrace()
            throw ApiException("Error while creating a secure connection.", e)
        } catch (e: SocketTimeoutException) {
            e.printStackTrace()
            throw TimeoutApiException("Connection error: " + e.message, e)
        } catch (e: InterruptedIOException) {
            e.printStackTrace()
            throw TimeoutApiException("Connection error: " + e.message, e)
        } catch (e: IOException) {
            e.printStackTrace()
            throw ApiException("Connection error: " + e.message, e)
        }
        val safe_url = request.url.toString().replace("^(.*)key=([0-9A-Za-z]+)([^0-9A-Za-z]*)".toRegex(), "$1key=redacted$3")
        sentry.addHttpBreadcrumb(safe_url, request.method, response.code)
        var body = ""
        if (json) {
            body = try {
                response.body!!.string()
            } catch (e: IOException) {
                e.printStackTrace()
                throw ApiException("Connection error: " + e.message, e)
            }
        }
        if (response.code >= 500) {
            response.close()
            throw ApiException("Server error: " + response.code)
        } else if (response.code == 404 && (!json || !body.startsWith("{"))) {
            response.close()
            throw NotFoundApiException("Server error: Resource not found.")
        } else if (response.code == 304) {
            throw ResourceNotModified()
        } else if (response.code == 403) {
            response.close()
            throw PermissionDeniedApiException("Server error: Permission denied.")
        } else if (response.code == 409) {
            response.close()
            throw ConflictApiException("Server error: " + response.code + ": " + body)
        } else if (response.code >= 405) {
            response.close()
            throw FinalApiException("Server error: " + response.code + ".")
        }
        if (response.code == 401) {
            if (body.startsWith("{")) {
                try {
                    val err = JSONObject(body)
                    if (err.optString("detail", "") == "Device access has been revoked.") {
                        throw DeviceAccessRevokedException("Device access has been revoked.")
                    }
                } catch (e: JSONException) {
                    e.printStackTrace()
                }
            }
        }
        return try {
            if (json && response.code != 204) {
                if (body.startsWith("[")) {
                    body = "{\"content\": $body}"
                }
                ApiResponse(
                        JSONObject(body),
                        response
                )
            } else {
                ApiResponse(
                        null,
                        response
                )
            }
        } catch (e: JSONException) {
            e.printStackTrace()
            sentry.captureException(e)
            throw ApiException("Invalid JSON received: " + body.take(100), e)
        }
    }

    @Throws(ApiException::class)
    open fun putFile(url: String, file: File, mediaType: MediaType, filename: String) {
        var request = Request.Builder()
                .url(url)
                .put(file.asRequestBody(mediaType))
                .header("Content-Disposition", "attachment; filename=\"$filename\"")
                .header("Authorization", "Device $key")
        if (acceptLanguage != null) {
            request = request.header("Accept-Language", acceptLanguage)
        }
        try {
            val resp = apiCall(request.build())
            if (resp.response.code != 201 && resp.response.code != 200) {
                throw FinalApiException("Could not upload file: " + resp.data)
            }
        } catch (exc: ResourceNotModified) { // can't happen
            throw FinalApiException("resource not modified")
        } catch (exc: JSONException) {
            throw FinalApiException("JSONException")
        }
    }

    @Throws(ApiException::class)
    open fun uploadFile(file: File, mediaType: MediaType, filename: String): String {
        var request = Request.Builder()
                .url(apiURL("upload"))
                .post(file.asRequestBody(mediaType))
                .header("Content-Disposition", "attachment; filename=\"$filename\"")
                .header("Authorization", "Device $key")
        if (acceptLanguage != null) {
            request = request.header("Accept-Language", acceptLanguage)
        }
        return try {
            val resp = apiCall(request.build())
            if (resp.response.code != 201) {
                throw FinalApiException("Could not upload file: " + resp.data)
            }
            resp.data!!.getString("id")
        } catch (exc: ResourceNotModified) { // can't happen
            throw FinalApiException("resource not modified")
        } catch (exc: JSONException) {
            throw FinalApiException("JSONException")
        }
    }

    companion object {
        /**
         * See https://docs.pretix.eu/en/latest/api/index.html for API documentation
         */
        const val SUPPORTED_API_VERSION = 4
        val JSON: MediaType = "application/json; charset=utf-8".toMediaType()
        fun fromConfig(config: ConfigStore, httpClientFactory: HttpClientFactory?=null, acceptLanguage: String? = null): PretixApi {
            return PretixApi(config.apiUrl, config.apiKey, config.organizerSlug,
                    config.apiVersion, httpClientFactory ?: DefaultHttpClientFactory(), acceptLanguage)
        }

        fun fromConfig(config: ConfigStore): PretixApi {
            return fromConfig(config, null)
        }
    }

    init {
        var url = url
        if (!url.endsWith("/")) {
            url += "/"
        }
        this.url = url
        this.key = key
        this.orgaSlug = orgaSlug
        this.version = version
        client = httpClientFactory.buildClient(NetUtils.ignoreSSLforURL(url))
        sentry = DummySentryImplementation()
    }
}
